# SI 618 - Homework #1: Data Manipulation

## Background

This homework assignment focuses on the analysis of historical data from the Olympic games.  The description of the data includes the following:
> This is a historical dataset on the modern Olympic Games, including all the Games from Athens 1896 to Rio 2016. I scraped this data from www.sports-reference.com in May 2018.

Your main task in this assignment is to explore the data *using the data
manipulation methods we covered in class* as well as those in the assigned readings.  You may need to consult pandas documentation, Stack Overflow, or other online resources.  

** You should also feel free to ask questions on the class Slack channel! **

A total of 100 points is available in this homework assignment, consisting of:
- 80 points for completing the specific homework questions. More comprehensive 
answers will tend to gain more points.
- 10 points for the overall quality of spelling, grammar, punctuation, and style of written responses.  (see https://faculty.washington.edu/heagerty/Courses/b572/public/StrunkWhite.pdf for a detailed specifications).
- 10 points for creating code that conforms to [PEP 8](https://www.python.org/dev/peps/pep-0008/) guidelines.  You should review those guidelines before proceding with the assignment.


## 0. Download the data from: 
- https://www.kaggle.com/heesoo37/120-years-of-olympic-history-athletes-and-results
- https://en.wikipedia.org/wiki/List_of_Olympic_Games_host_cities (note: we suggest you use the [read_html](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_html.html) function to load the data into a DataFrame.

In [2]:
MY_UNIQNAME = 'tengdann' # please fill in your uniqname

## 1. Answer the questions below. 
For each question, you should
1. Write code using Python and pandas that can help you answer the following questions, and
2. Explain your answers in plain English. You should use complete sentences that would be understood by an educated professional who is not necessarily a data scientist (like a product manager).

### 5 pt: Describe the dataset. 
- How many number rows and columns do you see in each data frame? 
- Are there any other interesting descriptive details?

In [64]:
import pandas as pd
%matplotlib inline

# Above libraries imported as per https://johnricco.github.io/2017/04/04/python-html/ to parse merged cells in html tables.

athletes = pd.read_csv('data/hw1_data/athlete_events.csv', index_col = None)
print(athletes.shape)
athletes.describe(include = 'all').fillna('')

(271116, 15)


ID                  Name     Sex      Age   Height   Weight  \
count    271116                271116  271116   261642   210945   208241   
unique                         134732       2                              
top              Robert Tait McKenzie       M                              
freq                               58  196594                              
mean      68249                                25.5569  175.339  70.7024   
std     39022.3                                6.39356  10.5185   14.348   
min           1                                     10      127       25   
25%       34643                                     21      168       60   
50%       68205                                     24      175       70   
75%      102097                                     28      183       79   
max      135571                                     97      226      214   

                 Team     NOC        Games     Year  Season    City  \
count          271116  271116       271116   271116  271116  271116   
unique           1184     230           51                2      42   
top     United States     USA  2000 Summer           Summer  London   
freq            17847   18853        13821           222552   22426   
mean                                        1978.38                   
std                                         29.8776                   
min                                            1896                   
25%                                            1960                   
50%                                            1988                   
75%                                            2002                   
max                                            2016                   

            Sport                    Event  Medal  
count      271116                   271116  39783  
unique         66                      765      3  
top     Athletics  Football Men's Football   Gold  
freq        38624                     5733  13372  
mean                                               
std                                                
min                                                
25%                                                
50%                                                
75%                                                
max

The athletes DataFrame has 271,116 rows and 15 columns.

The age of the athletes ranged from 10 to 97, with the median being 24.

In [63]:
regions = pd.read_csv('data/hw1_data/noc_regions.csv', index_col = None)
print(regions.shape)
regions.describe(include = 'all')

(230, 3)


NOC   region                     notes
count   230      227                        21
unique  230      206                        21
top     ISV  Germany  Turks and Caicos Islands
freq      1        4                         1

The regions DataFrame has 230 rows and 3 columns

### 5 pt: What is the overall ratio of female to male atheletes? 
There are multiple ways one could determine the meaning of "overall" here, and there isn't necessarily one correct answer. Based on your interpretation, give a justification for the choices you made in your analysis.

In [57]:
len(athletes[athletes['Sex'] == 'F']) / len(athletes[athletes['Sex'] == 'M'])

0.37906548521318045

The overall ratio of females to males is 0.379. I counted duplicate athletes, since competing multiple times adds to the total amount of athletes who ever competed.

### 10 pt: Which athelete competed in most number of events? 
In which events did the athelete participate, and for what range of years? Which country did the athlete represent?

In [6]:
athletes.Name.describe()

count                   271116
unique                  134732
top       Robert Tait McKenzie
freq                        58
Name: Name, dtype: object

In [7]:
tait = athletes[athletes.Name == 'Robert Tait McKenzie']
tait.Event.unique().tolist()

['Art Competitions Mixed Sculpturing',
 'Art Competitions Mixed Sculpturing, Statues',
 'Art Competitions Mixed Sculpturing, Medals And Reliefs',
 'Art Competitions Mixed Sculpturing, Unknown Event',
 'Art Competitions Mixed Sculpturing, Medals']

In [8]:
tait.Year.unique().tolist()

[1912, 1928, 1932, 1936, 1948]

In [9]:
tait.Team.unique().tolist()

['Canada']

The athlete who competed the most is Robert Tait McKenzie, who competed 58 times.  
He participated in:  
    1. Art Competitions Mixed Sculpturing  
    2. Art Competitions Mixed Sculpturing, Statues  
    3. Art Competitions Mixed Sculpturing, Medals and Reliefs  
    4. Art Competitions Mixed Sculpturing, Unknown Event  
    5. Art Competitions Mixed Sculpturing, Medals  
He competed from 1912 to 1948, representing Canada

### 10 pt: Which sport has the highest median athlete age? 
To determine this, you might want to plot the median age for each sport. You may want to sort the axes appropriately for your analysis.

In [59]:
athletes.groupby('Sport').Age.median().sort_values(ascending = False).head(5)

Sport
Roque               59.0
Art Competitions    45.0
Alpinism            38.0
Polo                35.0
Shooting            33.0
Name: Age, dtype: float64

Roque has the highest median athlete age.

### 10 pt: In which events has the Netherlands won the most medals?

In [12]:
netherlands = athletes[athletes.Team == 'Netherlands']
ned_medals = netherlands[pd.notnull(netherlands.Medal)]
ned_medals.groupby('Event').Medal.count().sort_values(ascending = False).head(5)

Event
Hockey Women's Hockey                              128
Hockey Men's Hockey                                127
Swimming Women's 4 x 100 metres Freestyle Relay     55
Rowing Women's Coxed Eights                         45
Football Men's Football                             41
Name: Medal, dtype: int64

The Netherlands has won the most medals in Women's Hockey (128 medals). I included all the multiple medals won by a team, instead of counting each year as 1 medal won by a team.

### 10 pt: Which countries have more female medalists than male medalists?

In [16]:
medalists = athletes[pd.notnull(athletes.Medal)].groupby('NOC').Sex.describe()
medalists_new = pd.merge(medalists, regions, on = 'NOC')
medalists_new.loc[medalists_new['top'] == 'F'].fillna('')

NOC  count  unique top  freq         region notes
14   BLR    139       2   F    77        Belarus      
18   BRN      3       1   F     3        Bahrain      
22   CHN    989       2   F   654          China      
26   CRC      4       1   F     4     Costa Rica      
68   JAM    157       2   F    86        Jamaica      
75   KOS      1       1   F     1         Kosovo      
80   LIE      9       2   F     5  Liechtenstein      
89   MNE     14       1   F    14     Montenegro      
91   MOZ      2       1   F     2     Mozambique      
103  PER     15       2   F    12           Peru      
107  PRK     67       2   F    34    North Korea      
110  ROU    653       2   F   355        Romania      
115  SGP      9       2   F     7      Singapore      
138  UKR    199       2   F   105        Ukraine      
148  ZIM     22       1   F    22       Zimbabwe

Belarus, Bahrain, China, Costa Rica, Jamaica, Kosovo, Liechtenstein, Montenegro, Mozambique, Peru, North Korea, Romania, Singapore, Ukraine, and Zimbabwe all have more female medalists than male medalists.

I also included countries that only have female athletes; inherently any positive integer is greater than 0. Since some countries technically sent 0 male athletes, they would fulfill the conditionn laid out by the question.

### 10 pt: Which country won the most medal points in each of the 2016 Summer Olympics?
Rank each country by a score, where 3 points are assigned for a gold medal, 2 points for silver, and 1 point for bronze.

In [61]:
def points(row):
    if row['Medal'] == 'Gold':
        val = 3
    elif row['Medal'] == 'Silver':
        val = 2
    elif row['Medal'] == 'Bronze':
        val = 1 
    else:
        val = 0
    return val

athletes['Medal_val'] = athletes.apply(points, axis = 1)
athletes[athletes.Games == '2016 Summer'].groupby('NOC').Medal_val.sum().sort_values(ascending = False).head(5)

NOC
USA    596
GBR    328
GER    300
RUS    247
CHN    235
Name: Medal_val, dtype: int64

The United States won the most medal points (596 points) in 2016 Summer Olympics.

### 20 pt: For each year in which games were held, what proportion of gold medals did the host country win? 
You will need to combine multiple datasets to complete the analysis.  

In [144]:
# Scrubbing functions and lines were created with the help of SI 618 study group

def clean_table(row):
    if row['City'][-1] == ']':
        row['City'] = row['City'][0:-3]
        
    return row['City']

link = 'https://en.wikipedia.org/wiki/List_of_Olympic_Games_host_cities'
host_cities = pd.read_html(link)[1] # God bless pandas 0.24.0
host_cities.City = host_cities.apply(clean_table, axis = 1)

# Need to manually adjust some of the country names
host_cities.Country = host_cities.Country.replace({'United Kingdom': 'England', 'Nazi Germany': 'Germany', 'Empire of Japan\xa0Nazi Germany': 'Japan', 'Empire of Japan\xa0Finland': 'Japan', 
                                   'Australia\xa0Sweden': 'Australia', 'West Germany': 'Germany', 'Soviet Union': 'Russia', 'Great Britain': 'England'})

athletes_copy = athletes.copy()

# Scrub the team names and bring them inline with host_cities country names
athletes_copy.Team = athletes_copy.Team.str.replace('-1', '').str.replace('2', '').str.replace('-3', '').str.replace('-4', '').str.replace('#1', '')
athletes_copy.Team = athletes_copy.Team.replace({'Nazi Germany': 'Germany', 'West Germany': 'Germany', 'Soviet Union': 'Russia', 'United Kingdom': 'England', 'Great Britain': 'England'})
athletes_copy.City = athletes_copy.City.replace({'Athina': 'Athens', 'Antwerpen': 'Antwerp', 'Sankt Moritz': 'St. Moritz'})

ath_copy_wiki = pd.merge(athletes_copy[['Team', 'Year', 'City', 'Medal']], host_cities[['City', 'Country', 'Year']], on = ['City', 'Year'])
ath_copy_wiki_filt = ath_copy_wiki[ath_copy_wiki.Medal == 'Gold']
gold_totals = ath_copy_wiki_filt.groupby('Year').count().Medal
gold_host = ath_copy_wiki_filt[ath_copy_wiki_filt.Team == ath_copy_wiki_filt.Country].groupby('Year').count().Medal
(gold_host / gold_totals).fillna(0)*100

Year
1896    16.129032
1900    10.945274
1904    38.728324
1908    31.292517
1912    29.141104
1920    10.953347
1924    11.144578
1928     7.272727
1932    29.501916
1936    27.298851
1948     2.121212
1952     4.273504
1956     3.921569
1960    38.000000
1964     9.068627
1968     1.647059
1972     9.493671
1976     0.393701
1980    33.333333
1984    32.574431
1988     4.612850
1992     7.993967
1994     7.272727
1996    26.151316
1998     5.517241
2000     9.049774
2002     5.555556
2004     1.204819
2008    11.028316
2010    38.505747
2012     7.594937
2014    16.336634
2016     5.413534
Name: Medal, dtype: float64

The above table contains the percent of gold medals won by the host country for every year the Olympics occured. I first made a copy of athletes, to ensure the original dataframe isn't altered.

hext came some filtering and editing: the wiki table had some interesting country names, so some research and collaboration with other class members led us to figure out what country the games were actually held in. T

Next came another merge with the pulled Wikipedia table, on city and year. This merged table was filtered by gold medals. Two separate new dataframes were created, one counting the total number of gold medals earned each year, and the other counting the number of gold medals the host country won. The percentage was calculated by dividing the latter by the former, and filling in NAs with 0.

There exists an issue here: if the noc_region.csv has a different NOC/region name than the athlete_events.csv/Wikipedia table, then the Python code would not match them. The only way is to correct this is to manually go through noc_region.csv and edit any region/NOC names that do not match. However, no dataset is always 100% clean, so even if it was manually altered, there may be some values that are missed.

## Please submit your completed notebook in .IPYNB and .HTML formats via Canvas